content/oci/oci.go

```diff
@@ -0,0 +1,228 @@
+	if reference == "" {
+		return errdef.ErrMissingReference
+	}
+
+	exists, err := s.storage.Exists(ctx, desc)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrNotFound)
+	}
+
+	// reference can be either a tag or a digest.
+	if isDigest(reference) {
+		// a digest reference must match the digest of the descriptor
+		if reference != desc.Digest.String() {
+			return fmt.Errorf("digest reference mismatch: %s: expect %s: %w", reference, desc.Digest.String(), errdef.ErrInvalidReference)
+		}
+	} else {
+		if desc.Annotations == nil {
+			desc.Annotations = map[string]string{}
+		}
+		desc.Annotations[ocispec.AnnotationRefName] = reference
+	}
+
+	s.resolver.Tag(ctx, desc, reference)
+	return s.saveIndex()
+}
+
+// Resolve resolves a reference to a descriptor.
+func (s *Store) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	if reference == "" {
+		return ocispec.Descriptor{}, errdef.ErrInvalidReference
+	}
+
+	return s.resolver.Resolve(ctx, reference)
+}
+
+func (s *Store) UpEdges(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	return s.finder.UpEdges(ctx, node)
+}
+
+// ensureOCILayoutFile ensures the `oci-layout` file.
+func (s *Store) ensureOCILayoutFile() error {
+	layoutFilePath := filepath.Join(s.root, ocispec.ImageLayoutFile)
+	layoutFile, err := os.Open(layoutFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+
+		layout := ocispec.ImageLayout{
+			Version: ocispec.ImageLayoutVersion,
+		}
+		layoutJSON, err := json.Marshal(layout)
+		if err != nil {
+			return err
+		}
+
+		return ioutil.WriteFile(layoutFilePath, layoutJSON, 0666)
+	}
+	defer layoutFile.Close()
+
+	var layout *ocispec.ImageLayout
+	err = json.NewDecoder(layoutFile).Decode(&layout)
+	if err != nil {
+		return err
+	}
+	if layout.Version != ocispec.ImageLayoutVersion {
+		return errdef.ErrUnsupportedVersion
+	}
+
+	return nil
+}
+
+// loadIndex reads the index.json from the file system.
+func (s *Store) loadIndex() error {
+	indexFile, err := os.Open(s.indexPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		s.index = &ocispec.Index{
+			Versioned: specs.Versioned{
+				SchemaVersion: 2, // historical value
+			},
+		}
+		return nil
+	}
+	defer indexFile.Close()
+
+	if err := json.NewDecoder(indexFile).Decode(&s.index); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	for _, desc := range s.index.Manifests {
+		// reference can be either a tag or a digest.
+		// tag by digest if tag is not present.
+		if desc.Annotations == nil || desc.Annotations[ocispec.AnnotationRefName] == "" {
+			s.resolver.Tag(ctx, desc, desc.Digest.String())
+		} else {
+			tag := desc.Annotations[ocispec.AnnotationRefName]
+			s.resolver.Tag(ctx, desc, tag)
+		}
+
+		// traverse the whole DAG and index up edges for all the nodes.
+		s.finder.IndexAll(ctx, s.storage, desc)
```

Comment:
It seems that the return values (error) of `Tag()` and `IndexAll()` are not checked.